matrix_client/api.py

```diff
@@ -582,16 +582,24 @@ def _send(self, method, path, content=None, query_params={}, headers={},
 
         if headers["Content-Type"] == "application/json" and content is not None:
             content = json.dumps(content)
+            
+        request_timeout = 30 + query_params.get("timeout", 30000) / 1000
 
         response = None
         while True:
-            response = requests.request(
-                method, endpoint,
-                params=query_params,
-                data=content,
-                headers=headers,
-                verify=self.validate_cert
-            )
+            try:
+                response = requests.request(
+                    method, endpoint,
+                    params=query_params,
+                    data=content,
+                    headers=headers,
+                    verify=self.validate_cert,
+                    timeout=request_timeout
+                )
+            except requests.exceptions.ReadTimeout:
+                raise MatrixTimeoutError(
```

Comment:
Pass the `requests` exception into the constructor for `MatrixTimeoutError`
so that people can debug easier. Also consider passing in the full URL that
was called and the method to be inserted into the error's `msg` (See
`MatrixRequestError` for an example of this).